### ISSUE TITLE
ArrayFireConfig.cmake does not work on windows

### ISSUE THREAD

**Issue by willyd** [arrayfire/arrayfire#925]
After installing ArrayFire v3. And setting ArrayFire_DIR to C:/Program Files/ArrayFire/v3/cmake. CMake complains that the library:

 C:/Program Files/lib/af${backend}/lib does not exist.

Removing 2 lines in ArrayFire${backend}.cmake fixes the problem:

    # Compute the installation prefix relative to this file.
    get_filename_component(_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
    get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
    # Commenting these lines fixes the problem.
    # get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
    # get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)

**Comment by pavanky** [comment]
Did you install ArrayFire from source or from the binary?

**Comment by willyd** [comment]
From the binary.

**Comment by pavanky** [comment]
For windows you should be able to use Visual Studio without using cmake. What is the situation you are using cmake in windows ?

**Comment by willyd** [comment]
I already have a project using CMake on Windows and wanted to try ArrayFire inside this project. I understand that I can specify the libraries by hand in VS but since the cmake config file are installed I suppose they should work.

**Comment by pavanky** [comment]
Fair enough. We will look into resolving this for 3.1

**Comment by willyd** [comment]
Thank you very much.

**Comment by shehzan10** [comment]
I personally don't think the cmake files that get shipped with the installer do a particularly good job of finding ArrayFire on windows. These files are auto generated and tweaking them might be more work than is worth. IMOI, the best solution is to ship a FindArrayFire.cmake file (custom for windows) that users like yourself can copy into your source code and use it easily.

Eventually, I think we can add a request to CMake to ship the FindArrayFire.cmake file that would be cross platform with the cmake installer.

**Comment by ghisvail** [comment]
Isn't it the purpose of the CMake package registry and [`export`](http://www.cmake.org/cmake/help/v2.8.12/cmake.html#command%3aexport) function for automatic detection on Windows ?

**Comment by shehzan10** [comment]
I'm not sure about how CMake detects things on windows. From the closest comparison, FindCUDA.cmake, it is a simple file that anyone can copy and include in their project. I have not seen any *-config.cmake files on windows (like the ones we ship with linux). I think it's better to ship a simple FindArrayFire.cmake file with the installer.
Alternatively, we can write a universal FindArrayFire.cmake file and host it on github/elsewhere. This should work across all platforms/OS and should be simple enough for users to include in their projects.

**Comment by shehzan10** [comment]
I created this file https://gist.github.com/shehzan10/1c41396aafb4a843c191 (not tested). But I think this should do the trick.
I simply gets AF_PATH from the environment and then sets everything up. I'll try setting up some test for this.

**Comment by shehzan10** [comment]
Update: I tested the file on windows and it works perfectly. Had to fix a typo but thats it.

I used @bkloppenborg's https://github.com/bkloppenborg/arrayfire-cmake-example to test this. I simply added the FindArrayFire.cmake file to the CMakeModules directory and it worked like a charm.

**Comment by bkloppenborg** [comment]
Humm. The `FindArrayFire.cmake` script shouldn't be needed if the `ArrayFireConfig.cmake` script is working correctly as they do essentially the same thing.

**Comment by shehzan10** [comment]
I agree. But I don't know if they work (and how they work) on windows. I've never had a good experience with those on windows.

**Comment by willyd** [comment]
I beleive that `PackageConfig.cmake` files are the recommended (or more modern) approach to package resolution on any platform with CMake enabled libraries. With the targets exporting feature, they make it easy for a package to expose its requirements to consuming projecs: preprocessor definitions, include directories, libraries. This is also doable with a find package module but much harder in my opinion. If a library is built using CMake it should generate a `PackageConfig.cmake`. 

From the CMake documention:

> A config-file package is a set of files provided by upstreams for downstreams to use. CMake searches in a number of locations for package configuration files, as described in the find_package() documentation. The most simple way for a CMake user to tell cmake(1) to search in a non-standard prefix for a package is to set the CMAKE_PREFIX_PATH cache variable.

You only need to install all package in some prefix folder and add that as an environment variable called `CMAKE_PREFIX_PATH`. Otherwise, you can do a first configure and wait for CMake to create the Package_DIR (ArrayFire_DIR in this case) variable and change the path to the folder conrtaining the `PackageConfig.cmake` file.

**Comment by pavanky** [comment]
@willyd I agree. We will not be dropping `ArrayFireConfig.cmake` 

We are trying to resolve the original issue before 3.1

**Comment by pavanky** [comment]
Having said that, @shehzan10's solution can be used as a work around until we fix the issue.

**Comment by ghisvail** [comment]
Isn't that what the whole package registry thing is about on Windows:

http://www.cmake.org/cmake/help/v3.0/manual/cmake-packages.7.html

**Comment by shehzan10** [comment]
@ghisvail Thanks for pointing that out. I was not aware of that. Let me experiment  and if successful, we'll add it to the installer.

**Comment by shehzan10** [comment]
So there are 2 parts to this:

1. The registry thing that @ghisvail mentioned is to allow automatic detection of the ArrayFire install directory. I have added this to our installer.

2. TL;DR: Fixed the original issue. Should be in the 3.1 installer.
Upon a little bit of digging (and actually reading the comment), those lines are to move from the CMake config files directory to the installation directory. When we were building the installer, we had the cmake files at package/share/ArrayFire/cmake (default), where "package" was CMake_Install_prefix. Hence to go from cmake to package, it wanted to go up 3 levels. Whereas, when we shipped the installer, the cmake directory was install_dir/cmake. Hence, removing 2 lines made it work.
I have fixed this in our build systems. I will be testing this installer soon and these improvements should be included in the 3.1 installer.
